src/cachemachine/rubinimageinfo.py

```diff
@@ -0,0 +1,346 @@
+import re
+from dataclasses import dataclass
+from enum import Enum, auto
+from typing import Optional
+
+import semver
+from docker_image import reference
+
+from .types import IncomparableImagesError
+
+DEFAULT_TAG = "latest"  # Docker convention
+DEFAULT_RECOMMENDED_TAG = "recommended"
+
+
+class RubinImageType(Enum):
+    """Class to capture the different kinds of image types possible in a
+    Rubin Lab Image container, given our standard tag set."""
+
+    RECOMMENDED = auto()
+    LATEST = auto()
+    LATEST_SOMETYPE = auto()
+    RELEASE = auto()
+    WEEKLY = auto()
+    DAILY = auto()
+    RELEASE_CANDIDATE = auto()
+    EXPERIMENTAL = auto()
+    MALFORMED_TAG = auto()
+
+
+# Helper functions that don't rely on any part of the class object and thus
+# are not methods.
+
+
+def _check_tag_is_lowercase(tag: str) -> bool:
+    if tag != tag.lower():
+        return False
+    return True
+
+
+def _type_from_tag(tag: str, recommended_tag: str) -> RubinImageType:
+    """This constructs the image type from the tag.  It is inherently
+    Rubin tag-specific."""
+    if not _check_tag_is_lowercase(tag):
+        return RubinImageType.MALFORMED_TAG
+    if tag == recommended_tag:
+        return RubinImageType.RECOMMENDED
+    if tag == DEFAULT_TAG:
+        return RubinImageType.LATEST
+    if tag.startswith("latest_"):
+        # It's latest *something*
+        ttag = tag[7:]
+        if ttag in ("weekly", "daily", "release"):
+            return RubinImageType.LATEST_SOMETYPE
+        return RubinImageType.MALFORMED_TAG
+    if tag.startswith("w_"):
+        return RubinImageType.WEEKLY
+    if tag.startswith("d_"):
+        return RubinImageType.DAILY
+    if tag.startswith("exp_"):
+        return RubinImageType.EXPERIMENTAL
+    if tag.startswith("r"):
+        # If there are any non-numeric characters in the tag after the inital
+        # 'r', it's a release candidate; if not, we claim it's a real release.
+        # Thus, r22_0_0_rc1 is a release candidate, but r21_0_1 is a release.
+        tagtext = tag[1:].replace("_", "")
+        if tagtext.isdigit():
+            return RubinImageType.RELEASE
+        return RubinImageType.RELEASE_CANDIDATE
+    return RubinImageType.MALFORMED_TAG
+
+
+def _semver_from_tag(tag_type: RubinImageType, tag: str) -> semver.VersionInfo:
+    if (
+        tag_type == RubinImageType.EXPERIMENTAL
+        or tag_type == RubinImageType.RECOMMENDED
+        or tag_type == RubinImageType.LATEST
+        or tag_type == RubinImageType.LATEST_SOMETYPE
+        or tag_type == RubinImageType.MALFORMED_TAG
+    ):
+        # The "latest" tags are by their nature singletons.
+        # Experimentals and malformed tags do not have meaningful
+        # semantic versioning and are compared purely lexigraphically by tag.
+        return semver.VersionInfo(0, 0, 1)
```

Comment:
This seems really strange, shouldn't this function just return an error instead of a bogus semver?  It seems like we shouldn't be comparing semver and non semver things

Reply:
Well, it shouldn't return an error, since it's what I am using to populate the semantic_version field for a RubinImageInfo object, but making that field Optional[semver.VersionInfo] and putting None in there would be fine with me.